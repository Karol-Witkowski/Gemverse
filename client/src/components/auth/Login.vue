<template>
  <v-container class="mt-12">
    <v-card
      class="mt-12 mx-auto"
      max-width="500px"
    >
      <v-card-title>
        <span class="headline grey--text text--darken-2">Sign in to start chatting *disabled*</span>
      </v-card-title>
      <v-card-text>
        <v-container>
          <v-form
            ref="form"
            v-model="isFormValid"
          >
            <v-row>
              <v-col cols="12">
                <v-text-field
                  label="Username"
                  required
                  :rules="generalRules.concat(usernameRules)"
                  v-model="username"
                />
              </v-col>
              <v-col cols="12">
                <v-text-field
                  label="Password"
                  required
                  :rules="generalRules.concat(passwordRules)"
                  type="password"
                  v-model="password"
                />
              </v-col>
            </v-row>
            <span>All fields are case-sensitive</span>
          </v-form>
        </v-container>
      </v-card-text>
      <v-card-actions class="pb-4">
        <v-btn
          color="primary"
          text
          to="/"
          outlined
        >
          Close
        </v-btn>
        <v-spacer />
        <v-btn
          color="primary"
          :disabled="!isFormValid"
          text
          to="/roomlist"
          outlined
        >
          Sign in
        </v-btn>
      </v-card-actions>
    </v-card>
    <div class="mt-6 text-center">
      <h4 class="grey--text mx-auto mb-2 text--darken-1">New to Gemverse? Create a new account</h4>
      <v-btn
        color="blue lighten-1"
        text
        to="/register"
        outlined
      >
        Sign up
      </v-btn>
    </div>
  </v-container>
</template>

<script>
export default {
  name: 'Login',
  data() {
    return {
      password: '',
      username: '',
      isFormValid: false,
      usernameRules: [
        (value) => value.length <= 128 || 'Username must be less or equal to 128 characters',
      ],
      passwordRules: [
        (value) => value.length <= 128 || 'Password must be less or equal to 128 characters',
      ],
      generalRules: [
        (value) => !!value || 'Required',
      ],
    };
  },
};
</script>
