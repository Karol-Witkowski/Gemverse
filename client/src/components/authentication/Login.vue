<template>
  <v-container class="mt-12">
    <v-alert
      class="mt-8 mx-auto"
      dense
      max-width="500"
      type="error"
      v-bind:class="[authError ? 'authErrorAlert' : 'whiteSpace']"
    >
      <v-row align="center">
        <v-col class="grow">
          <strong class="mx-auto">{{ authError }}</strong>
        </v-col>
        <v-col class="shrink">
          <v-btn
            @click="showAuthError"
            small
          >
            OK
          </v-btn>
        </v-col>
      </v-row>
    </v-alert>
    <v-card
      class="mt-8 mx-auto"
      max-width="500px"
    >
      <v-card-title>
        <span class="headline grey--text text--darken-2">Sign in to start chatting</span>
      </v-card-title>
      <v-card-text>
        <v-container>
          <v-form
            ref="form"
            v-model="isFormValid"
          >
            <v-row>
              <v-col cols="12">
                <v-text-field
                  autofocus
                  :error-messages="userError"
                  label="E-mail adress"
                  required
                  :rules="generalRules.concat(emailRules)"
                  v-model="email"
                  v-on:keyup="[userError = '', passwordError = '']"
                  v-on:keyup.enter="formValidation"
                />
              </v-col>
              <v-col cols="12">
                <v-text-field
                  :error-messages="passwordError"
                  label="Password"
                  required
                  :rules="generalRules.concat(passwordRules)"
                  type="password"
                  v-model="password"
                  v-on:keyup="passwordError = ''"
                  v-on:keyup.enter="formValidation"
                />
              </v-col>
            </v-row>
            <span>All fields are case-sensitive</span>
          </v-form>
        </v-container>
      </v-card-text>
      <v-card-actions class="pb-4">
        <v-btn
          color="primary"
          text
          to="/"
          outlined
        >
          Back
        </v-btn>
        <v-spacer />
        <v-btn
          @click.prevent="formValidation"
          color="primary"
          :disabled="!isFormValid"
          text
          outlined
        >
          Sign in
        </v-btn>
      </v-card-actions>
    </v-card>
    <div class="mt-6 text-center">
      <h4 class="grey--text mx-auto mb-2 text--darken-1">New to Gemverse? Create a new account</h4>
      <v-btn
        color="blue lighten-1"
        text
        to="/register"
        outlined
      >
        Sign up
      </v-btn>
    </div>
  </v-container>
</template>

<script>
import axios from 'axios';

export default {
  name: 'Login',
  props: ['message'],
  data() {
    return {
      authError: this.message,
      email: '',
      isFormValid: false,
      password: '',
      passwordError: '',
      placeholder: 'whiteSpace',
      userError: '',
      emailRules: [
        (value) => value.length <= 128 || 'E-mail adress must be less or equal to 128 characters',
      ],
      passwordRules: [
        (value) => value.length <= 128 || 'Password must be less or equal to 128 characters',
      ],
      generalRules: [
        (value) => !!value || 'Required',
      ],
    };
  },

  created() {
    this.setAuthToken();
  },

  methods: {
    dispatchToken() {
      if (localStorage.getItem('authenticationToken')) {
        this.$store.dispatch('remitAuthState', true);
      } else {
        localStorage.clear();
        this.$store.dispatch('remitAuthState', false);
      }
    },

    formValidation() {
      if (this.isFormValid) {
        this.login();
      }
    },

    login() {
      axios.post('http://localhost:3000/api/authentication/login', {
        email: this.email,
        password: this.password,
      })
        .then((response) => {
          localStorage.setItem('authenticationToken', response.data.token);
          this.setAuthToken(response.data.token);
          this.dispatchToken();

          if (response.status === 200) {
            this.$router.push({
              name: 'RoomList',
            });
          }
        })
        .catch((error) => {
          console.log(error);
          this.userError = error.response.data.user;
          this.passwordError = error.response.data.password;
        });
    },

    setAuthToken(token) {
      if (token) axios.defaults.headers.common.Authorization = token;
      else delete axios.defaults.headers.common.Authorization;
    },

    showAuthError() {
      this.authError = '';
    },
  },
};
</script>

<style>
.authErrorAlert {
  visibility: visible;
}

.whiteSpace {
  opacity: 0;
  transition: visibility 0.3s linear,opacity 0.3s linear;
  visibility: hidden;
}
</style>
