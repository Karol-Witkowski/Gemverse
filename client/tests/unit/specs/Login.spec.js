import { createLocalVue, mount } from '@vue/test-utils';
import axios from 'axios';
import Vue from 'vue';
import Vuetify from 'vuetify';
import Login from '@/components/authentication/Login.vue';

const mockStore = { dispatch: jest.fn() };
const localVue = createLocalVue();
const url = 'http://localhost:3000/api/authentication/login';
let vuetify;
let wrapper;
let expectedData = expect.objectContaining({
  email: 'email value',
  password: 'password value',
});

jest.mock('axios');
vuetify = new Vuetify();

describe('Implementation test for Login.vue - successful HTTP post', () => {
  let response = {
    data: {
      auth:	true,
      data: {
        data: 'testData'
      },
      success: true,
      token: "testToken",
    },
  };

  beforeEach(() => {
    axios.post.mockResolvedValue(response);

    wrapper = mount(Login, {
      localVue,
      mocks: {
        $store: mockStore
      },
      stubs: ['router-link', 'router-view'],
      vuetify,
      data() {
        return {
          email: '',
          isFormValid: false,
          password: '',
          redirectError: this.message,
        };
      },
      propsData: {
        message: '',
      },
    });
  });

  afterEach(() => {
    axios.post.mockReset();
    wrapper.destroy();
  });

  it('Render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Initializes with correct elements', () => {
    // Test buttons initial state
    expect(wrapper.findAll('.v-btn').length).toEqual(4);
    expect(wrapper.findAll('.v-btn').at(0).text()).toMatch('ok');
    expect(wrapper.findAll('.v-btn').at(1).text()).toMatch('back');
    expect(wrapper.findAll('.v-btn').at(2).text()).toMatch('sign in');
    expect(wrapper.findAll('.v-btn').at(3).text()).toMatch('sign up');
    expect(wrapper.findAll('.v-btn').at(2).element.disabled).toBeTruthy();

    // Test auth alert initial state
    expect(wrapper.findAll('.v-alert').at(0).attributes().class).toContain('whiteSpace');

    // Test inputs initial state
    expect(wrapper.findAll('.v-text-field').length).toEqual(2);
    expect(wrapper.findAll('input').at(0).text()).toEqual('');
    expect(wrapper.findAll('input').at(1).text()).toEqual('');

    // Test validation initial state
    expect(wrapper.findAll('.v-messages').length).toEqual(2);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('');
    expect(wrapper.findAll('.v-messages').at(1).text()).toEqual('');
    expect(wrapper.vm.isformValid).toBeFalsy();
  });

  it('Enables log-in button when email address and password are entered', async () => {
    await wrapper.setData({
      email: 'email value',
      password: 'password value',
    });

    await Vue.nextTick();

    // Check that the user data is properly set
    expect(wrapper.findAll('input').at(0).element.value).toEqual('email value');
    expect(wrapper.findAll('input').at(1).element.value).toEqual('password value');
    expect(wrapper.vm.email).toBe('email value');
    expect(wrapper.vm.password).toBe('password value');

    // Check that the login button is active
    expect(wrapper.findAll('.v-btn').at(2).element.disabled).toBeFalsy();
  });

  it('Pass validation when email and password are entered', async () => {
    await wrapper.setData({
      email: 'email value',
      password: 'password value',
    });

    await Vue.nextTick();

    expect(wrapper.vm.isFormValid).toBeTruthy();
  });

  it('Fail validation when email and password are not entered', async () => {
    expect(wrapper.vm.isFormValid).toBeFalsy();
  });

  it('Fail validation when one field is empty', async () => {
    await wrapper.setData({
      email: 'email value',
    });

    await Vue.nextTick();

    expect(wrapper.vm.isFormValid).toBeFalsy();
  });

  it('Enables auth alert when authentication failed', async () => {
    await wrapper.setData({
      redirectError: 'Access denied',
    });

    await Vue.nextTick();

    expect(wrapper.findAll('.v-alert').at(0).attributes().class).toContain('errorAlert');
  });

  it('Should sends post request with correct on form submit', async () => {
    await wrapper.setData({
      email: 'email value',
      password: 'password value',
    });

    await Vue.nextTick();

    wrapper.vm.login();

    // Check if post is called
    expect(axios.post).toHaveBeenCalled();

    // Check if post are called once
    expect(axios.post).toHaveReturnedTimes(1);

    // Check if post is called with correct data
    expect(axios.post).toHaveBeenCalledWith(
      url,
      expectedData,
    );
  });

  it('Should store user data and auth status after successful login', async () => {
    wrapper.vm.login();

    // Check if any action are dispatched
    expect(mockStore.dispatch).toHaveBeenCalled();

    // Check if two actions are dispatched
    expect(mockStore.dispatch).toHaveReturnedTimes(2);

    // Check auth state is dispatched with correct data
    expect(mockStore.dispatch).toHaveBeenNthCalledWith(
      1, 'remitAuthState', true
    );

    // Check user data is dispatched with correct data
    expect(mockStore.dispatch).toHaveBeenNthCalledWith(
      2, 'saveUser', response.data.data,
    );
  });
});

describe('Implementation test for Login.vue - failed HTTP post', () => {
  beforeEach(() => {
    axios.post.mockRejectedValue(new Error('BAD REQUEST'));

    wrapper = mount(Login, {
      localVue,
      mocks: {
        $store: mockStore
      },
      stubs: ['router-link', 'router-view'],
      vuetify,
      data() {
        return {
          email: '',
          isFormValid: false,
          password: '',
        };
      },
    });
  });

  afterEach(() => {
    axios.post.mockReset();
    mockStore.dispatch.mockReset();
    wrapper.destroy();
  });

  it('Call login function with correct data', async () => {
    await wrapper.setData({
      email: 'email value',
      password: 'password value',
    });

    await Vue.nextTick();

    wrapper.vm.login();

    expect(axios.post).toHaveBeenCalledTimes(1);
    expect(axios.post).toBeCalledWith(
      url,
      expectedData
    );
  });

  it('Does not dispatch data when a failed HTTP post occurs', () => {
    wrapper.vm.login();

    expect(mockStore.dispatch).not.toHaveBeenCalled();
  });
});
