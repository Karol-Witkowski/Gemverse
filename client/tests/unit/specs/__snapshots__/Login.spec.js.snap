// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Implementation test for Login.vue - successful HTTP post Render correctly 1`] = `
<div class="container mt-12">
  <div role="alert" class="v-alert mt-8 mx-auto v-sheet theme--dark v-alert--dense error whiteSpace" style="max-width: 500px;">
    <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark"></i>
      <div class="v-alert__content">
        <div class="row align-center">
          <div class="grow col"><strong class="mx-auto"></strong></div>
          <div class="shrink col"><button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--dark v-size--small"><span class="v-btn__content">
          ok
        </span></button></div>
        </div>
      </div>
    </div>
  </div>
  <div class="mt-8 mx-auto v-card v-sheet theme--light" style="max-width: 500px;">
    <div class="v-card__title"><span class="headline grey--text text--darken-2">Sign in to start chatting</span></div>
    <div class="v-card__text">
      <div class="container">
        <form novalidate="novalidate" class="v-form">
          <div class="row">
            <div class="col col-12">
              <div class="v-input theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot"><label for="input-7" class="v-label theme--light" style="left: 0px; position: absolute;">E-mail adress</label><input required="required" autofocus="autofocus" id="input-7" type="text"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div class="col col-12">
              <div class="v-input theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot"><label for="input-10" class="v-label theme--light" style="left: 0px; position: absolute;">Password</label><input required="required" id="input-10" type="password"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div> <span>All fields are case-sensitive</span>
        </form>
      </div>
    </div>
    <div class="v-card__actions pb-4">
      <router-link-stub class="v-btn v-btn--outlined v-btn--router v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">
        back
      </span></router-link-stub>
      <div class="spacer"></div> <button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--outlined v-btn--text theme--light v-size--default"><span class="v-btn__content">
        sign in
      </span></button>
    </div>
  </div>
  <div class="mt-6 text-center">
    <h4 class="grey--text mb-2 mx-auto text--darken-1">New to Gemverse? Create a new account</h4>
    <router-link-stub class="v-btn v-btn--outlined v-btn--router v-btn--text theme--light v-size--default blue--text text--lighten-1"><span class="v-btn__content">
      sign up
    </span></router-link-stub>
  </div>
</div>
`;
